build-system/server.js

```diff
@@ -99,12 +120,32 @@ app.use('/form/html/post', function(req, res) {
 });
 
 app.use('/form/echo-json/post', function(req, res) {
+  if(!ORIGIN_REGEX.test(req.headers.origin)) {
+    res.statusCode = 500;
+    res.end(JSON.stringify({
+      message: 'Origin header is invalid.'
+    }));
+    return;
+  }
+
+  if(!SOURCE_ORIGIN_REGEX.test(req.query.__amp_source_origin)) {
+    res.statusCode = 500;
+    res.end(JSON.stringify({
+      message: '__amp_source_origin parameter is invalid.'
+    }));
+    return;
+  }
+
   var form = new formidable.IncomingForm();
   form.parse(req, function(err, fields) {
     res.setHeader('Content-Type', 'application/json');
     if (fields['email'] == 'already@subscribed.com') {
       res.statusCode = 500;
     }
+    res.setHeader('Access-Control-Allow-Origin',
+        req.protocol + '://' + req.headers.host);
+    res.setHeader('Access-Control-Expose-Headers',
+        'AMP-Access-Control-Allow-Source-Origin')
     res.setHeader('AMP-Access-Control-Allow-Source-Origin',
         req.protocol + '://' + req.headers.host);
```

Comment:
Technically this is wrong. You should just echo `req.query.__amp_source_origin` here.